[Woo POS] Use `Observation` to fix refresh issue

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #14869
Merge after: #15059 
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR removes the `.value` workaround for the refresh control on the item lists, without re-introducing the error shown when you pull to refresh (PTR).

It does so by reducing the scope of redraws from item list state changes to just the `ItemListView` and `ChildItemList`, then moving the `refreshable` declaration to where we call the view.

This is required because we do change the item list state every time we pull to refresh – we can improve this further by not setting the state to `loading` when we pull to refresh, but I want to do that separately to keep things clear. This is just a note for info – setting the state to `loading` means the ghost cell shows at the bottom of the list during PTR... but that's not the right loading indicator. Instead, we can rely on the built-in loading indicator which shows at the top of the list, and avoid changing the content at all. When we do that, we can move the `refreshable` modifier back to the list views.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app and open POS
2. Wait for the list to load
3. Pull to refresh
4. Observe that it works as expected, and no error is shown.
5. Repeat with a variation list

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

This PR makes changes to the state handling of the item lists – instead of making immutable copies, we now use mutable classes to hold our state, as is the way with `Observation`, which relies on `willSet` to publish changes.

For that reason, it's worth checking pagination, errors, initial loading of the two different item lists.

Unit testing intermediate states from the Observable gets pretty hairy. The `loading` states in the tests here are OK-ish, because it's just the first state that comes back after we start watching. When we need the second or third, but not final, intermediate state, we need to juggle re-observation in the `withObservationTracking` call. It doesn't help that they use `willSet` not `didSet` semantics.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/c70e835a-41c6-47aa-9900-34c6185e2807

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
14869 Remove workaround for refresh cancellation

joshheald committed:
14869 Make items controller a nested observable

joshheald committed:
14869 Make itemsViewState’s properties mutable

joshheald committed:
14869 Make ItemsViewState observable

joshheald committed:
14869 Make ItemsStackState observable

joshheald committed:
14869 Move refresh control to fix refresh issue

Now only the ItemListView and ChildItemList are redrawn when we pull to refresh, not the dashboard.

This means that we can avoid redrawing the `refreshable` modifier by moving it up the heirarchy, to where the views are used.

Moving it up means that we can continue to use the state in the view itself, without cancelling the refresh task as soon as it starts, and consequently showing an error.

This is only needed because we change the list state when we pull to refresh, to `.loading`, which should arguably be reserved for loading the next page (as it adds the ghost cell at the end of the list)

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Fix lint issue

joshheald committed:
Merge branch 'issue/14869-make-pos-aggregate-model-observable' into issue/14869-use-observation-to-fix-refresh-issue

joshheald committed:
14869 Remove Copiable due to iOS 17 requirement

CodeGen can’t add `@available` headers to its generated code, so the next time someone generates new `GeneratedCopiable` implementation files, these `copy` functions for the item list would break.

joshheald committed:
Merge branch 'issue/14869-use-observation-to-fix-refresh-issue' of github.com:woocommerce/woocommerce-ios into issue/14869-use-observation-to-fix-refresh-issue

joshheald committed:
14869 Update test doubles to require iOS 17

joshheald committed:
14869 update intermediate tests to use Observation

joshheald commented:
@staskus @iamgabrielma @jaclync only one review required, just adding you all for visibility.

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -49,9 +50,7 @@ protocol PointOfSaleAggregateModelProtocol {
-    private(set) var itemsViewState: ItemsViewState = ItemsViewState(containerState: .loading,
-                                                                     itemsStack: ItemsStackState(root: .loading([]),
-                                                                                                 itemStates: [:]))
+    var itemsViewState: ItemsViewState { itemsController.itemsViewState }
```
Ha, it really is that simple to pass the state with `Observation` 👍 I was curious how we'll achieve that.

In your experimentations, does applying any mappings here break any of the observation behavior? Or we will be really able to replace most of the publishers this way?

joshheald replied:
I think it'll work for most things – Observation works automatically with any computed property which is based on other observable properties.

If you have a computed property with no reliance on other observables, you need to write your own registration code but I've not had to do that yet.

staskus commented on WooCommerce/WooCommerceTests/POS/Controllers/PointOfSaleItemsControllerTests.swift:
```diff
@@ -1,26 +1,18 @@
-        sut.itemsViewStatePublisher.assign(to: &$itemsViewState)
-    }
-
+    @available(iOS 17.0, *)
```
I assume it's not possible to put it at a class level?

joshheald replied:
lol no (╯°□°）╯︵ ┻━┻

staskus commented on WooCommerce/WooCommerceTests/POS/Controllers/PointOfSaleItemsControllerTests.swift:
```diff
@@ -424,18 +429,22 @@ final class PointOfSaleItemsControllerTests {
-            Issue.record("Expected loading ItemList state, but got \(sut.itemsViewState)")
-            return
+        await confirmation() { confirmation in
+            withObservationTracking {
```
TIL

staskus commented on WooCommerce/WooCommerceTests/POS/Controllers/PointOfSaleItemsControllerTests.swift:
```diff
@@ -331,74 +394,88 @@ final class PointOfSaleItemsControllerTests {
+            } onChange: {
+                // Then
+                Task { @MainActor in
+                    // Dispatched in a task because `onChange` is called with `willSet` semantics 🙄
```
Okay, that's annoying. I tried to look for ways to avoid this but I haven't found anything good 🤔 

I haven't written a `confirmation()` before. It was a bit counterintuitive for me that it doesn't wait for the result of `await` and immediately proceeds to load the items. However, it fits the needs since we need to confirm an intermediate loading valu.e

joshheald replied:
It's just like the old `wait(for expectation:)` that we use.

It waits for all the `await`s, but the `await confirmation` one is at a higher scope than the `await loadItems` one. Note where the braces are.

1. So it starts waiting for the `confirmation()` call, which doesn't finish until we check for the loading state. 
2. Within that scope, it sets up the observation tracking so we can do the check, then calls `await loadItems()` which it also waits for.
3. Before loadItems returns, we `#expect` the loading state, and the confirmation is called so it won't auto-fail when the `await confirmation` returns.
4. Then `loadItems()` returns, but we don't assert anything here
5. Then `await confirmation()` returns and the test ends.

staskus reviewed (commented):
Thanks! Amazing. Great to see the items state simplified by using `@Observable`

> Observe that it works as expected, and no error is shown.

Unfortunately, I'm getting errors by playing around with pull to refresh. iPad Air M2 18.2

https://github.com/user-attachments/assets/819dc119-0caf-40dc-8e28-8bd2d43cc90f

joshheald commented:
Thanks for the review @staskus 

> Unfortunately, I'm getting errors by playing around with pull to refresh. iPad Air M2 18.2
> ScreenRecording_02-05-2025.23-58-25_1.MP4

Hmm. So, any change to state while the view's refreshing cancels the refresh, which actually makes sense here – when you scroll to the bottom and start a new page load, it doesn't make sense to continue with the refresh. The error you see is that the refresh call was cancelled.

We need to swallow those cancellation errors, because they're not relevant to the user. Doing that in this PR would make it harder to tell whether or not we've fixed the problem, so I'll do it in a followup if that's OK with you?

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/237">21.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented:
> We need to swallow those cancellation errors, because they're not relevant to the user. Doing that in this PR would make it harder to tell whether or not we've fixed the problem, so I'll do it in a followup if that's OK with you?

Thanks for the explanation! That totally makes sense, @joshheald !

joshheald committed:
Merge branch 'trunk' into issue/14869-use-observation-to-fix-refresh-issue

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/Item Selector/ChildItemList.swift
#	WooCommerce/Classes/POS/Presentation/ItemListView.swift

joshheald committed:
Fix merge issue

joshheald committed:
Merge branch 'trunk' into issue/14869-use-observation-to-fix-refresh-issue
